skimage/util/_backends.py

```diff
@@ -0,0 +1,153 @@
+    searched for an implementation. If no backend implements the function
+    then the scikit-image implementation is used.
+    """
+    func_name = func.__name__
+    func_module = get_module_name(func)
+
+    # If no backends are installed at all or dispatching is disabled,
+    # return the original function. This way people who don't care about it
+    # don't see anything related to dispatching
+    if disable_dispatching() or not all_backends():
+        return func
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        # Backends are tried in alphabetical order, this makes things
+        # predictable and stable across runs. Might need a better solution
+        # when it becomes common that users have more than one backend
+        # that would accept a call.
+        for name in sorted(all_backends()):
+            backend = all_backends()[name]
+            # If we have a BackendInformation object we check if the
+            # function we are looking for is implemented in the backend
+            if "info" in backend:
+                if (
+                    f"{func_module}:{func_name}"
+                    not in backend["info"].supported_functions
+                ):
+                    continue
+
+            backend_impl = backend["implementation"].load()
+
+            # Allow the backend to accept/reject a call based on the function
+            # name and the values of the arguments
+            wants_it = backend_impl.can_has(
+                f"{func_module}:{func_name}", *args, **kwargs
+            )
+            if not wants_it:
+                continue
+
+            # can_has("foo", ...) might be True, but the function might not actually
+            # be implemented in the backend
+            func_impl = backend_impl.get_implementation(f"{func_module}:{func_name}")
+            if func_impl is not None:
+                warnings.warn(
+                    f"Call to '{func_module}:{func_name}' was dispatched to"
+                    f" the '{name}' backend. Set SKIMAGE_NO_DISPATCHING=1 to"
+                    " disable this.",
+                    DispatchNotification,
+                    # XXX from where should this warning originate?
+                    # XXX from where the function that was dispatched was called?
+                    # XXX or from where the user called a function that called
+                    # XXX a function that was dispatched?
+                    stacklevel=2,
+                )
+                return func_impl(*args, **kwargs)
+
+        else:
+            return func(*args, **kwargs)
+
+    return wrapper
+
+
+class BackendInformation:
+    """Information about a backend
+
+    A backend that wants to provide additional information about itself
+    should return an instance of this from its information entry-point.
+    """
+
+    def __init__(self, supported_functions):
+        self.supported_functions = supported_functions
```

Comment:
Using an abstract base class might be reasonable here. E.g.

```python
from abc import ABCMeta, abstractmethod

class ABCBackendInformation(metaclass=ABCMeta):
    @property
    @abstractmethod
    def supported_functions(self):
        ...
```

If we require importing this "interface" above, then we should also check it with `isinstance`. Otherwise why not remove this and just rely on ducktyping?

Reply:
Why make it an abstract base class? Backends should use this class as is, not inherit from it or some such.

For now this only contains the list of implemented functions, but I was thinking we could also use it to hold other information like the fancy/pretty name of the backend, the URL to its documentation and hence JSON doc, etc.

By providing a class backends can use, we hopefully make it easy for them to provide the information we need. We could just use a `dict` instead of course but it seems like using a (data)class is nicer/more in line with what people are doing?

---

While thinking about this I did realise that this class is in `skimage.util._backends` which signals that you should not use it from outside scikit-image. We could expose it at `skimage.util`? Or tell backend authors that "this is fine for you to use, but it isn't part of the public API and hence comes with less backwards compatibility promises"? WDYT